File: python/leetcode/152.py
class Solution(object):
    def maxProduct(self, nums):
        if not nums:
            return 0
        vmax, vmin, res = nums[0], nums[0], nums[0]
        for i in range(1, len(nums)):
            if nums[i] < 0:
                vmax, vmin = vmin, vmax
            vmax = max(nums[i], nums[i] * vmax)
            vmin = min(nums[i], nums[i] * vmin)
            res = max(res, vmax)
        return res

File: cpp/leetcode/152.cpp
class Solution {
public:
    int maxProduct(vector<int> &nums) {
        if (nums.empty()) {
            return 0;
        }
        int pmin = nums[0], pmax = nums[0], res = nums[0];
        for (int i = 1; i < nums.size(); i++) {
            if (nums[i] < 0) {
                swap(pmin, pmax);
            }
            pmax = max(pmax * nums[i], nums[i]);
            pmin = min(pmin * nums[i], nums[i]);
            res = max(res, pmax);
        }
        return res;
    }
};